sdk/identity/azure-identity/inc/azure/identity/azure_cli_credential.hpp

```diff
@@ -112,7 +112,14 @@ namespace Azure { namespace Identity {
   protected:
 #endif
     virtual std::string GetAzCommand(std::string const& scopes, std::string const& tenantId) const;
-    virtual int GetLocalTimeToUtcDiffSeconds() const;
+
+#if !defined(TESTING_BUILD)
+  private:
+#else
+  protected:
+    virtual
+#endif
```

Comment:
Preprocessor conditional virtuals will generate to bugs, which are hard to diagnose and debug.
If customer had `TESTING_BUILD` defined when building the library, but does not have it defined when consuming from their code, vtbl indexes will be different. The code will behave randomly and will eventually crash, with very strange error and callstack.
Conditional virtuals create more problems than they fix.
And we tested with Larry that for virtuals in final class, the code calls them as non-virtual, so there is no perf hit.
I think Larry has agreed to that, the issue you are referencing should have been just closed without any PR?

Reply:
> If customer had `TESTING_BUILD` defined when building the library, but does not have it defined when consuming from their code, vtbl indexes will be different. The code will behave randomly and will eventually crash, with very strange error and callstack.

This is **not** a supported scenario and hence we shouldn't be catering to this. If we have customer feedback that they want to build with testing on (and I don't see why they would), we can address this at that time.

Reverting this change is easier to do in the future, than the other way around.

Reply:
It is easier to get into this without knowing than you think.
`TESTING_BUILD` is defined when building the SDK. It is not defined when the customer is consuming our code.
And it's not like the customer is setting `TESTING_BUILD` themselves, when they build the SDK. They set `BUILD_TESTING` CMake flag (or is it even on by default?). Nowhere in the docs we say that the binaries that were built with `BUILD_TESTING` CMake flag on should not be consumed by an outside code.

Reply:
Our CI process builds the SDK with both BUILD_TESTING on and off. I know, because I've had to deal with build breaks because of it (similarly, building perf tests are unrelated to building other tests (strangely, stress tests are always built with other tests).

However, the entire *existence* of a the TESTING_BUILD macro configuration is problematic - because it means we *never* run tests on the code our customers use in the configuration our customers use.

That's why *every* place we use TESTING_BUILD should be looked at with a great deal of suspicion.